File: JavaTestClient/src/com/dreamcrushed/DeviceUser/Packet/Packet.java
/**
 * Pennyworth - A new smarthome protocol.
 * Copyright (C) 2012  Dream-Crusher Labs LLC
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.dreamcrushed.DeviceUser.Packet;


import java.lang.reflect.InvocationTargetException;

import com.dreamcrushed.DeviceUser.InvalidPacketException;
import com.dreamcrushed.DeviceUser.NetworkManager;

/**
 * This is the basic unit of communication between the TerraSoul
 * server and client. Classes extend packet and are designed to
 * be self handling through the handle method. They also have
 * unique states where they are transferred to a string and back
 * for network transfer implemented using the packetString and
 * newPacket methods.
 * 
 * @author jmonk
 */
public abstract class Packet {

	public Packet(String[] params) {
	}

	/**
	 * Creates an instance of a packet based on the type contained in the
	 * packet string.
	 * 
	 * @param s Packet String
	 * @return Packet
	 */
	public static Packet newPacket(String s) throws InvalidPacketException {
		String[] split = s.split(":");
		PacketType type = PacketType.fromInt(Integer.parseInt(split[0]));
		if (type == null) {
			throw new InvalidPacketException();
		}
		String[] params = new String[split.length - 2];
		for (int i = 2; i < split.length; i++) {
			params[i - 2] = split[i];
		}
		try {
			return type.instance(Integer.parseInt(split[1]), params);
		} catch (NumberFormatException e) {
			throw new InvalidPacketException();
		} catch (SecurityException e) {
			throw new InvalidPacketException();
		} catch (NoSuchMethodException e) {
			throw new InvalidPacketException();
		} catch (IllegalArgumentException e) {
			throw new InvalidPacketException();
		} catch (InstantiationException e) {
			throw new InvalidPacketException();
		} catch (IllegalAccessException e) {
			throw new InvalidPacketException();
		} catch (InvocationTargetException e) {
			e.getTargetException().printStackTrace();
			throw new InvalidPacketException();
		}
	}

	/**
	 * Handles any client side parsing required by this packet. Could be anything
	 * from sending the user a message to modifying health of a character on screen.
	 * Should be overloaded in any packet which is to do anything.
	 *
	 * @param networkManager
	 */
	public void handle(NetworkManager networkManager) {
		networkManager.removePacket(this);
	}

	/**
	 * Turn this packet into a string that can be transferred over the network.
	 * 
	 * @return Packet String
	 */
	public abstract String packetString();

}

File: Server/DevicePackets/Packet.cpp
/**
 * Pennyworth - A new smarthome protocol.
 * Copyright (C) 2012  Dream-Crusher Labs LLC
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/*
 * Packet.cpp
 *
 *  Created on: Jun 18, 2012
 *      Author: jmonk
 */
#include "Packet.h"
#include <stdio.h>
#include <stdlib.h>
namespace dvs {

Communicator* currentCommunicator = NULL;

Packet::Packet(unsigned char *data) {
//	printf("Data Packet %d\n", data[3]);
	this->data = data;
	this->length = data[0] + 1;
	this->device = (((unsigned int) data[2]) << 8) | data[1];
	allocated = false;
}
Packet::Packet(unsigned int length, unsigned int device, unsigned char pid) {
//	printf("Other Packet %d %d\n", pid, device);
	this->length = length;
	this->device = device;
	this->data = (unsigned char *) malloc(sizeof(unsigned char) * length);
	this->pid = pid;
	if (this->data) {
		this->data[0] = length - 1;
		this->data[1] = device & 0xff;
		this->data[2] = (device >> 8) & 0xff;
		this->data[3] = pid;
		this->allocated = false;
	} else {
		perror("Memory Allocation Failed:");
		this->data = NULL;
		this->length = 0;
		this->allocated = false;
	}
}

Packet::~Packet() {
	if (allocated) {
		free(data);
		data = NULL;
	}
}

inline Device* Packet::getDevice() {
	return Device::getDevice(device);
}

void Packet::send() {
	getDevice()->send(this);
}

unsigned char Packet::getType() {
	return pid;
}

unsigned char *Packet::getData() {
	return data;
}

unsigned int Packet::getLength() {
	return length;
}

} /* namespace DVS */
